tests/stub/versions/test_versions.py

```diff
@@ -20,21 +25,73 @@ def tearDown(self):
-        result = session.run("RETURN 1 AS n")
-        # Otherwise the script will not fail when the protocol is not present
-        # (on backends where run is lazily evaluated)
-        result.next()
-        session.close()
-        driver.close()
+        try:
+            yield session
+        finally:
+            session.close()
+            driver.close()
+            self._server.reset()
+
+    def _run(self, version, server_agent=None, check_version=False,
+             rejected_agent=False, check_server_address=False):
+        """
+            :param version: e.g. "4x3" or "3"
+            :type version: string
+            :param server_agent: Set a specific server agent and check that the
+                                 summery list the correct one.
+            :type server_agent: str, optional
+            :param check_version: Check if the summery contains the correct bolt
+                                  version
+            :type check_version: bool
+            :param rejected_agent:
+                True: Expect the driver to fail by rejecting the server agent
+                False: Expect the driver not to fail
+            :type rejected_agent: bool
+            :param check_server_address: Check if the summery contains the right
+                                         server-address
+            :type check_server_address: bool
+        """
+        expected_version = version.replace("x", ".")
+        if "." not in expected_version:
+            expected_version += ".0"
+        vars_ = {}
+        if server_agent is None:
+            vars_["#SERVER_AGENT#"] = "Neo4j/" + expected_version + ".0"
+        else:
+            vars_["#SERVER_AGENT#"] = server_agent
+        script_path = self.script_path("v{}_return_1.script".format(version))
+        with self._get_session(script_path, vars_=vars_) as session:
+            try:
+                result = session.run("RETURN 1 AS n")
+                if server_agent or check_version or check_server_address:
+                    summary = result.consume()
+                    if check_version:
+                        self.assertEqual(summary.server_info.protocol_version,
+                                         expected_version)
+                    if server_agent is not None:
+                        self.assertEqual(summary.server_info.agent,
+                                         vars_["#SERVER_AGENT#"])
+                    if check_server_address:
+                        self.assertEqual(summary.server_info.address,
+                                         self._server.address)
+                else:
+                    # Otherwise the script will not fail when the protocol is
+                    # not present (on backends where run is lazily evaluated)
+                    result.next()
+            except DriverError:
+                if not rejected_agent:
+                    raise
+                # TODO: check exception further
```

Comment:
Are you sure you want to keep this comment?

Reply:
I think, once we agree on a common exception hierarchy/naming and add testkit messages for them, this test should be extended to check for the correct error to be thrown.